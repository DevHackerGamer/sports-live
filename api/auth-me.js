const { getUserType } = require('../lib/auth');

// Unified handler: resolves user type via Clerk when available, else allows dev header override
module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-User-Type');

  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'GET') return res.status(405).json({ success: false, error: 'Method not allowed' });

  try {
    const type = (await getUserType(req)) || 'user';
    const isAdmin = String(type).toLowerCase() === 'admin';
    return res.status(200).json({ success: true, type, isAdmin });
  } catch (error) {
    console.error('auth-me error:', error);
    return res.status(200).json({ success: true, type: 'user', isAdmin: false });
  }
};
