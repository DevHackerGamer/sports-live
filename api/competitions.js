const { getMatchesCollection } = require('../lib/mongodb');

module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'GET') return res.status(405).json({ success:false, error:'Method not allowed'});
  try {
    const matchesCollection = await getMatchesCollection();
    const names = await matchesCollection.distinct('competition.name', { 'competition.name': { $exists: true } });
    const filtered = names.filter(n => typeof n === 'string' && n.trim()).sort();
    res.status(200).json({ success:true, data: filtered, count: filtered.length });
  } catch (e) {
    res.status(500).json({ success:false, error:'Failed to load competitions', message: e.message });
  }
};
