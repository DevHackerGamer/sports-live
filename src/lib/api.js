// API client for making HTTP requests to our MongoDB-based backend
// This replaces Firebase real-time database functionality

class ApiClient {
  constructor(baseUrl = '') {
    this.baseUrl = baseUrl;
  }

  async request(endpoint, options = {}) {
    const url = `${this.baseUrl}${endpoint}`;
    const config = {
      headers: {
        'Content-Type': 'application/json',
  // Optional dev pass-through; in production backend should validate Clerk token
  ...(options.userType ? { 'X-User-Type': options.userType } : {}),
        ...options.headers,
      },
      ...options,
    };

    if (config.body && typeof config.body === 'object') {
      config.body = JSON.stringify(config.body);
    }

    try {
      const response = await fetch(url, config);
      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || `HTTP ${response.status}`);
      }

      return data;
    } catch (error) {
      console.error(`API request failed: ${url}`, error);
      throw error;
    }
  }

  // Matches API
  async getMatches() {
  // Request a broader upcoming window explicitly
  return this.request('/api/matches?range=10&includePast=&limit=500');
  }

  async getMatchesByDate(dateFrom, dateTo, limit = 1000) {
    const params = new URLSearchParams({ dateFrom, dateTo, limit: String(limit) });
    return this.request(`/api/matches?${params.toString()}`);
  }

  async getMatch(id) {
    return this.request(`/api/matches/${id}`);
  }
    // Alias for compatibility
    async getMatchById(id) {
      return this.getMatch(id);
    }
  
  async getMatchEvents(id) {
    return this.request(`/api/matches/${id}/events`);
  }
  
  async addMatchEvent(id, event, options = {}) {
    return this.request(`/api/matches/${id}/events`, {
      method: 'POST',
      body: event,
      ...options,
    });
  }
  
  async updateMatchEvent(id, eventId, updates, options = {}) {
    return this.request(`/api/matches/${id}/events/${eventId}`, {
      method: 'PUT',
      body: updates,
      ...options,
    });
  }
  
  async deleteMatchEvent(id, eventId, options = {}) {
    return this.request(`/api/matches/${id}/events/${eventId}`, {
      method: 'DELETE',
      ...options,
    });
  }
  
  // Update a match (accepts optional fetch options like headers)
  async updateMatch(id, updates, options = {}) {
    return this.request(`/api/matches/${id}`, {
      method: 'PUT',
      body: updates,
      ...options,
    });
  }

  // Get all news or by league
  async getFootballNews(leagueCode = null, limit = 20) {
    let url = `/api/football-news?limit=${limit}`;
    if (leagueCode) url += `&leagueCode=${leagueCode}`;
    return this.request(url);
  }
// Get football highlights
async getFootballHighlights(leagueName) {
  return this.request(`/api/football-highlights?leagueName=${encodeURIComponent(leagueName)}`);
}


  async createMatches(matches) {
    return this.request('/api/matches', {
      method: 'POST',
      body: { matches },
    });
  }

  // (Legacy duplicate removed â€“ single updateMatch above handles this)

  async deleteMatch(id) {
    return this.request(`/api/matches/${id}`, {
      method: 'DELETE',
    });
  }

  // Match Statistics API
  async getMatchStatistics(matchId) {
    return this.request(`/api/match-statistics?matchId=${encodeURIComponent(matchId)}`);
  }

  async createMatchStatistics(matchId, statistics) {
    return this.request('/api/match-statistics', {
      method: 'POST',
      body: { matchId, ...statistics },
    });
  }

  async updateMatchStatistics(matchId, statistics) {
    return this.request('/api/match-statistics', {
      method: 'PUT',
      body: { matchId, ...statistics },
    });
  }

  async deleteMatchStatistics(matchId) {
    return this.request(`/api/match-statistics?matchId=${encodeURIComponent(matchId)}`, {
      method: 'DELETE',
    });
  }

  // Reports API - recognize change`
async getReports(id) {
  if (id) {
    return this.request(`/api/reporting/${id}`);
  }
  return this.request('/api/reporting');
}

async createReport(report) {
  return this.request('/api/reporting', {
    method: 'POST',
    body: report,
  });
}

async updateReport(id, updates) {
  return this.request(`/api/reporting/${id}`, {
    method: 'PUT',
    body: updates,
  });
}

async deleteReport(id) {
  return this.request(`/api/reporting/${id}`, {
    method: 'DELETE',
  });
}


  // Teams API
  async getTeams() {
    return this.request('/api/teams');
  }

  async createTeams(teams) {
    return this.request('/api/teams', {
      method: 'POST',
      body: { teams },
    });
  }
  async getTeamById(id) {
  return this.request(`/api/teams?id=${id}`);
}

  //standings API

  // Standings API
    async getStandings({ competition, season, type, stage, limit = 20 } = {}) {
    const params = new URLSearchParams();
    if (competition) params.append('competition', competition);
    if (season) params.append('season', season);
    if (type) params.append('type', type);
    if (stage) params.append('stage', stage);
    params.append('limit', limit);

    return this.request(`/api/standings?${params.toString()}`);
  }

  async getStandingById(id) {
    return this.request(`/api/standings/${id}`);
  }

  // Users API
  async getUserFavorites(userId) {
    return this.request(`/api/users/${userId}/favorites`);
  }

  async addUserFavorite(userId, teamName) {
    return this.request(`/api/users/${userId}/favorites`, {
      method: 'POST',
      body: { teamName },
    });
  }

  async removeUserFavorite(userId, teamName) {
  const encoded = encodeURIComponent(teamName);
  return this.request(`/api/users/${userId}/favorites/${encoded}`, {
      method: 'DELETE',
    });
  }

  async updateUserFavorites(userId, favorites) {
    return this.request(`/api/users/${userId}/favorites`, {
      method: 'PUT',
      body: { favorites },
    });
  }

  async getEventLog(params = {}) {
    const qs = new URLSearchParams();
    if (params.limit) qs.append('limit', String(params.limit));
    if (params.type) qs.append('type', params.type);
    if (params.matchId) qs.append('matchId', params.matchId);
    if (params.startDate) qs.append('startDate', params.startDate);
    if (params.endDate) qs.append('endDate', params.endDate);
    const query = qs.toString();
    return this.request(`/api/event-log${query ? '?' + query : ''}`);
  }

  // User Watchlist (Matches)
  async getUserWatchlist(userId) {
    const params = new URLSearchParams();
    if (userId) params.append('userId', userId);
    return this.request(`/api/user-matches?${params.toString()}`);
  }

  async addUserMatch(userId, match) {
    const matchId = match?.id || match?._id || match?.matchId;
    if (!matchId) throw new Error('matchId required');
    return this.request('/api/user-matches', {
      method: 'POST',
      body: { userId, matchId, match },
    });
  }

  async removeUserMatch(userId, matchId) {
    const qs = new URLSearchParams({ userId, matchId: String(matchId) });
    return this.request(`/api/user-matches?${qs.toString()}`, { method: 'DELETE' });
  }
}

// Create a singleton instance
const apiClient = new ApiClient();

// Real-time data simulation using polling
class RealTimeData {
  constructor(apiClient) {
    this.apiClient = apiClient;
    this.subscriptions = new Map();
    this.intervals = new Map();
  }

  // Simulate Firebase onValue with polling
  onValue(endpoint, callback, errorCallback, pollInterval = 30000) {
    const id = Math.random().toString(36);
    
    // Initial fetch
    this.fetchAndNotify(endpoint, callback, errorCallback);
    
    // Set up polling
    const intervalId = setInterval(() => {
      this.fetchAndNotify(endpoint, callback, errorCallback);
    }, pollInterval);
    
    this.intervals.set(id, intervalId);
    this.subscriptions.set(id, { endpoint, callback, errorCallback });
    
    // Return unsubscribe function
    return () => {
      if (this.intervals.has(id)) {
        clearInterval(this.intervals.get(id));
        this.intervals.delete(id);
      }
      this.subscriptions.delete(id);
    };
  }

  async fetchAndNotify(endpoint, callback, errorCallback) {
    try {
      let data;
      switch (endpoint) {
        case 'matches':
          data = await this.apiClient.getMatches();
          // Convert array to object with IDs as keys (Firebase-like structure)
          const matchesObj = {};
          data.data.forEach(match => {
            matchesObj[match.id] = match;
          });
          callback({ val: () => matchesObj });
          break;
        case 'teams':
          data = await this.apiClient.getTeams();
          callback({ val: () => data.data });
          break;
        default:
          if (endpoint.startsWith('users/') && endpoint.includes('/favorites')) {
            const userId = endpoint.split('/')[1];
            data = await this.apiClient.getUserFavorites(userId);
            callback({ val: () => data.data });
          }
          break;
      }
    } catch (error) {
      console.error('Error fetching data:', error);
      if (errorCallback) {
        errorCallback(error);
      }
    }
  }

  // Clean up all subscriptions
  cleanup() {
    this.intervals.forEach(intervalId => clearInterval(intervalId));
    this.intervals.clear();
    this.subscriptions.clear();
  }
}

const realTimeData = new RealTimeData(apiClient);

// Firebase-compatible API
export const db = {
  // Placeholder for compatibility
};

export const ref = (db, path) => {
  return { path };
};

export const onValue = (ref, callback, errorCallback) => {
  const path = ref.path;
  return realTimeData.onValue(path, callback, errorCallback);
};

export const update = async (ref, data) => {
  const path = ref.path;
  
  if (path.startsWith('users/') && path.includes('/favorites')) {
    const userId = path.split('/')[1];
    await apiClient.updateUserFavorites(userId, data.favorites);
  } else if (path === 'matches') {
    // Handle bulk match updates
    const matchesArray = Object.values(data);
    await apiClient.createMatches(matchesArray);
  } else if (path === 'teams') {
    await apiClient.createTeams(data);
  }
};

export const set = async (ref, data) => {
  return update(ref, data);
};

export const get = async (ref) => {
  const path = ref.path;
  
  try {
    let data;
    switch (path) {
      case 'matches':
        data = await apiClient.getMatches();
        return { val: () => data.data };
      case 'teams':
        data = await apiClient.getTeams();
        return { val: () => data.data };
      default:
        if (path.startsWith('users/') && path.includes('/favorites')) {
          const userId = path.split('/')[1];
          data = await apiClient.getUserFavorites(userId);
          return { val: () => data.data };
        }
        return { val: () => null };
    }
  } catch (error) {
    console.error('Error getting data:', error);
    return { val: () => null };
  }
};

export const child = (ref, path) => {
  return { path: `${ref.path}/${path}` };
};

// Export API client for direct use
export { apiClient, realTimeData };
// Export getMatchById for direct use
export const getMatchById = (id) => apiClient.getMatchById(id);

export default {
  db,
  ref,
  onValue,
  update,
  set,
  get,
  child,
  apiClient,
  realTimeData,
};
