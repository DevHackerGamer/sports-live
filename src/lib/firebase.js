// Firebase client initialization (App + Realtime Database)
import { initializeApp, getApps, getApp } from 'firebase/app';
import { getAnalytics } from 'firebase/analytics';
import {
  getDatabase,
  ref as dbRef,
  update as dbUpdate,
  set as dbSet,
  child as dbChild,
  get as dbGet,
  onValue as dbOnValue  // <-- add this
} from 'firebase/database';

// Firebase web config (public by design)
const firebaseConfig = {
  apiKey: 'AIzaSyDzZ0CAodwCYHgQJahsgcxJrmWiIIbVGM',
  authDomain: 'sports-live-c25ca.firebaseapp.com',
  databaseURL: 'https://sports-live-c25ca-default-rtdb.firebaseio.com',
  projectId: 'sports-live-c25ca',
  storageBucket: 'sports-live-c25ca.firebasestorage.app',
  messagingSenderId: '532399482994',
  appId: '1:532399482994:web:3d1353393886b64dc1baeb',
  measurementId: 'G-8JT6X5DF6B',
};

// Singleton app instance across HMR/dev reloads
const app = getApps().length ? getApp() : initializeApp(firebaseConfig);

// Realtime Database instance
export const db = getDatabase(app);

// Modular functions for convenience
export const ref = dbRef;
export const update = dbUpdate;
export const set = dbSet;
export const child = dbChild;
export const get = dbGet;
export const onValue = dbOnValue;  // <-- export onValue

// Guard analytics for browser-only environments
export const initAnalytics = () => {
  if (typeof window !== 'undefined' && 'measurementId' in firebaseConfig) {
    try {
      return getAnalytics(app);
    } catch (_) {
      // ignore if analytics not supported
    }
  }
  return null;
};

export default app;
