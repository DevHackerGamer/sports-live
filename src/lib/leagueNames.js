// Centralized league name mapping for ESPN league codes
// This provides consistent display names across the entire application

export const LEAGUE_NAMES = {
  // ESPN league codes
  'eng.1': 'Premier League',
  'esp.1': 'La Liga',
  'ita.1': 'Serie A',
  'ger.1': 'Bundesliga',
  'fra.1': 'Ligue 1',
  'uefa.champions': 'Champions League',
  'ned.1': 'Eredivisie',
  'por.1': 'Primeira Liga',
  'eng.2': 'Championship',
  'bra.1': 'Campeonato Brasileiro SÃ©rie A',
  'conmebol.libertadores': 'Copa Libertadores',
  
  // Football-data.org codes (legacy)
  'PL': 'Premier League',
  'PD': 'La Liga',
  'SA': 'Serie A',
  'BL1': 'Bundesliga',
  'FL1': 'Ligue 1',
  'CL': 'Champions League',
  'DED': 'Eredivisie',
  'PPL': 'Primeira Liga',
  'ELC': 'Championship'
};

/**
 * Get display name for a league code
 * @param {string} code - ESPN league code (e.g., 'eng.1') or name with bracket notation (e.g., 'Bundesliga [ger.1]')
 * @returns {string} Display name (e.g., 'Premier League')
 */
export function getLeagueName(code) {
  if (!code) return 'Unknown League';
  
  // Handle strings with bracket notation like "Bundesliga [ger.1]"
  const bracketMatch = /^(.*?)\s*\[([^\]]+)\]$/.exec(String(code));
  if (bracketMatch) {
    const [, namePart, codePart] = bracketMatch;
    // Try to get the name from the code part first
    const nameFromCode = LEAGUE_NAMES[codePart];
    if (nameFromCode) return nameFromCode;
    // Otherwise return the name part without brackets
    return namePart.trim();
  }
  
  return LEAGUE_NAMES[code] || code;
}

/**
 * Get league code from display name or return as-is if already a code
 * @param {string} nameOrCode - League name or code
 * @returns {string} League code
 */
export function getLeagueCode(nameOrCode) {
  if (!nameOrCode) return '';
  
  // If it's already a code, return it
  if (LEAGUE_NAMES[nameOrCode]) return nameOrCode;
  
  // Find code by name
  const entry = Object.entries(LEAGUE_NAMES).find(([_, name]) => name === nameOrCode);
  return entry ? entry[0] : nameOrCode;
}

/**
 * Format competition object to ensure it has proper display name
 * @param {Object} competition - Competition object with name/code
 * @returns {Object} Competition with displayName
 */
export function formatCompetition(competition) {
  if (!competition) return { name: 'Unknown League', displayName: 'Unknown League' };
  
  const name = competition.name || competition.code || competition.id || '';
  const displayName = getLeagueName(name);
  
  return {
    ...competition,
    displayName,
    code: competition.code || name
  };
}
