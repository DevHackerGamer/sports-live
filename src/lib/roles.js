// Client-side role utilities to interpret Clerk metadata consistently

export function getUserRoles(user) {
  if (!user) return [];
  const raw = [
    user?.privateMetadata?.type,
    user?.publicMetadata?.type,

  ]
    .concat(Array.isArray(user?.privateMetadata?.roles) ? user.privateMetadata.roles : [])
    .concat(Array.isArray(user?.publicMetadata?.roles) ? user.publicMetadata.roles : [])
    .filter(Boolean)
    .map(v => String(v).toLowerCase());
  return Array.from(new Set(raw));
}

export function isAdminFromUser(user) {
  const roles = getUserRoles(user);
  return roles.includes('admin');
}
