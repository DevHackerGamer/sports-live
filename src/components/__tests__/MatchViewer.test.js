import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import MatchViewer from '../matchViewer/MatchViewer';
import * as ClerkReact from '@clerk/clerk-react';
import * as api from '../../lib/api';

// Mock Clerk
jest.mock('@clerk/clerk-react', () => ({
  useUser: jest.fn(),
}));

// Mock API client
jest.mock('../../lib/api', () => ({
  apiClient: {
    getMatchById: jest.fn(),
    getTeams: jest.fn(),
    createReport: jest.fn(),
    addMatchEvent: jest.fn(),
    updateMatch: jest.fn(),
  },
}));

// Mock child components
jest.mock('../liveInput/LiveInput', () => ({ isAdmin, match, onBackToMatch }) => (
  <div data-testid="live-input">
    Live Input - Admin: {isAdmin ? 'Yes' : 'No'} - Match: {match?.id}
    <button onClick={onBackToMatch}>Back to Match</button>
  </div>
));

jest.mock('../matchViewer/MatchStatistics', () => ({ match }) => (
  <div data-testid="match-statistics">
    Statistics for {match?.homeTeam} vs {match?.awayTeam}
  </div>
));

jest.mock('../matchViewer/LineupsTab', () => ({ match }) => (
  <div data-testid="lineups-tab">
    Lineups for {match?.homeTeam} vs {match?.awayTeam}
  </div>
));

jest.mock('../matchViewer/LiveCommentaryFeed', () => ({ matchId }) => (
  <div data-testid="live-commentary">
    Live Commentary for match {matchId}
  </div>
));

// Mock utilities
jest.mock('../../lib/leagueNames', () => ({
  getLeagueName: jest.fn((code) => code || 'Unknown League'),
}));

describe('MatchViewer Component', () => {
  const mockOnBack = jest.fn();
  const mockUser = { id: 'user123', firstName: 'John' };
  const mockAdminUser = { 
    id: 'admin123', 
    firstName: 'Admin',
    privateMetadata: { type: 'admin' }
  };

  const mockMatch = {
    id: 'match1',
    homeTeam: { name: 'Arsenal', crest: '/arsenal.png' },
    awayTeam: { name: 'Chelsea', crest: '/chelsea.png' },
    homeScore: 2,
    awayScore: 1,
    status: 'live',
    competition: { name: 'Premier League', code: 'PL' },
    utcDate: '2024-01-15T15:00:00Z',
    venue: 'Emirates Stadium',
    matchday: 1,
    referee: 'Michael Oliver',
    events: [
      {
        id: 'event1',
        type: 'goal',
        minute: 23,
        team: 'Arsenal',
        player: 'Bukayo Saka',
        description: 'Goal - Arsenal - Bukayo Saka',
      },
      
      {
        id: 'event2',
        type: 'yellow_card',
        minute: 45,
        team: 'Chelsea',
        player: 'Reece James',
        description: 'Yellow Card - Chelsea - Reece James',
      },
    ],
  };

  const mockTeamsData = {
    data: [
      { _id: 'team1', name: 'Arsenal', crest: '/arsenal.png' },
      { _id: 'team2', name: 'Chelsea', crest: '/chelsea.png' },
    ],
  };

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Default mocks
    ClerkReact.useUser.mockReturnValue({ user: mockUser });
    
    // Mock API responses
    api.apiClient.getMatchById.mockResolvedValue({ data: mockMatch });
    api.apiClient.getTeams.mockResolvedValue(mockTeamsData);
    api.apiClient.createReport.mockResolvedValue({});
    api.apiClient.addMatchEvent.mockResolvedValue({});
    api.apiClient.updateMatch.mockResolvedValue({});
    
    // Mock league names
    const { getLeagueName } = require('../../lib/leagueNames');
    getLeagueName.mockImplementation((code) => {
      const leagues = { PL: 'Premier League', UCL: 'Champions League' };
      return leagues[code] || code || 'Unknown League';
    });
  });

  it('renders loading state when only matchId is provided', () => {
    render(<MatchViewer matchId="match1" onBack={mockOnBack} />);

    expect(screen.getByText('Loading match details...')).toBeInTheDocument();
  });

  it('renders no match selected state', () => {
    render(<MatchViewer onBack={mockOnBack} />);

    expect(screen.getByText('Select a match from the Live Sports tab to view details')).toBeInTheDocument();
  });

  it('renders match details when match is provided', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    expect(screen.getByText('Match Overview')).toBeInTheDocument();
    expect(screen.getByText('Arsenal')).toBeInTheDocument();
    expect(screen.getByText('Chelsea')).toBeInTheDocument();
    expect(screen.getByText('2')).toBeInTheDocument(); // Home score
    expect(screen.getByText('1')).toBeInTheDocument(); // Away score
  });

  it('navigates between sections', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    // Default to details section
    expect(screen.getByText('Premier League')).toBeInTheDocument();

    // Navigate to statistics
    fireEvent.click(screen.getByText('Statistics'));
    expect(screen.getByTestId('match-statistics')).toBeInTheDocument();

    // Navigate to lineups
    fireEvent.click(screen.getByText('Lineups'));
    expect(screen.getByTestId('lineups-tab')).toBeInTheDocument();

    // Navigate to events
    fireEvent.click(screen.getByText('Event Timeline'));
    expect(screen.getByText('Match Events Timeline')).toBeInTheDocument();
  });

  it('shows admin sections for admin users', async () => {
    ClerkReact.useUser.mockReturnValue({ user: mockAdminUser });

    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    expect(screen.getByText('Update Events')).toBeInTheDocument();
    expect(screen.getByText('Live Input')).toBeInTheDocument();
  });

  it('hides admin sections for non-admin users', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    expect(screen.queryByText('Update Events')).not.toBeInTheDocument();
    expect(screen.queryByText('Live Input')).not.toBeInTheDocument();
  });

  it('fetches match details when matchId is provided', async () => {
    await act(async () => {
      render(<MatchViewer matchId="match1" onBack={mockOnBack} />);
    });

    await waitFor(() => {
      expect(api.apiClient.getMatchById).toHaveBeenCalledWith('match1');
    });
  });

  it('handles back button click', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    const backButton = screen.getByText('← Back');
    fireEvent.click(backButton);

    expect(mockOnBack).toHaveBeenCalled();
  });

  it('displays match events timeline', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Event Timeline'));

    expect(screen.getByText('Goal - Arsenal - Bukayo Saka')).toBeInTheDocument();
    expect(screen.getByText('Yellow Card - Chelsea - Reece James')).toBeInTheDocument();
  });

  it('shows and hides event report panel', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Event Timeline'));

    const reportButton = screen.getByText('Report an Issue');
    fireEvent.click(reportButton);

    expect(screen.getByText('Report Event Issue')).toBeInTheDocument();
    expect(screen.getByText('Hide Report Panel')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Hide Report Panel'));

    expect(screen.queryByText('Report Event Issue')).not.toBeInTheDocument();
  });

  it('submits event report', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Event Timeline'));
    fireEvent.click(screen.getByText('Report an Issue'));

    // Fill report form
    const titleInput = screen.getByPlaceholderText('Brief title of the issue');
    const descriptionInput = screen.getByPlaceholderText('Describe the issue...');

    fireEvent.change(titleInput, { target: { value: 'Test Report' } });
    fireEvent.change(descriptionInput, { target: { value: 'Test description' } });

    const submitButton = screen.getByText('Submit');
    await act(async () => {
      fireEvent.click(submitButton);
    });

    expect(api.apiClient.createReport).toHaveBeenCalledWith({
      matchId: 'match1',
      eventId: '',
      title: 'Test Report',
      description: 'Test description',
    });
  });

  it('handles live input section for admin', async () => {
    ClerkReact.useUser.mockReturnValue({ user: mockAdminUser });

    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Live Input'));

    expect(screen.getByTestId('live-input')).toBeInTheDocument();
    expect(screen.getByText('Back to Match')).toBeInTheDocument();
  });

  it('updates match metadata as admin', async () => {
    ClerkReact.useUser.mockReturnValue({ user: mockAdminUser });

    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Update Events'));

    const refereeInput = screen.getByPlaceholderText('Referee');
    const venueInput = screen.getByPlaceholderText('Venue');

    fireEvent.change(refereeInput, { target: { value: 'Anthony Taylor' } });
    fireEvent.change(venueInput, { target: { value: 'Stamford Bridge' } });

    const saveButton = screen.getByText('Save Match Info');
    await act(async () => {
      fireEvent.click(saveButton);
    });

    expect(api.apiClient.updateMatch).toHaveBeenCalledWith(
      'match1',
      { referee: 'Anthony Taylor', venue: 'Stamford Bridge' },
      { userType: 'admin' }
    );
  });

  it('adds new event as admin', async () => {
    ClerkReact.useUser.mockReturnValue({ user: mockAdminUser });

    // Mock players data
    global.fetch = jest.fn().mockResolvedValue({
      ok: true,
      json: () => Promise.resolve({ players: ['Bukayo Saka', 'Martin Ødegaard'] }),
    });

    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Update Events'));

    // Fill event form
    const timeInput = screen.getByPlaceholderText('Time (e.g., 23:45)');
    const teamSelect = screen.getByDisplayValue('Select Team');
    const typeSelect = screen.getByDisplayValue('Goal');
    const descriptionInput = screen.getByPlaceholderText('Event description');

    fireEvent.change(timeInput, { target: { value: '67:00' } });
    fireEvent.change(teamSelect, { target: { value: 'Arsenal' } });
    fireEvent.change(typeSelect, { target: { value: 'goal' } });
    fireEvent.change(descriptionInput, { target: { value: 'Great goal!' } });

    const addButton = screen.getByText('Add Event');
    await act(async () => {
      fireEvent.click(addButton);
    });

    expect(api.apiClient.addMatchEvent).toHaveBeenCalledWith(
      'match1',
      {
        type: 'goal',
        time: '67:00',
        player: '',
        team: 'Arsenal',
        description: 'Great goal!',
      },
      { userType: 'admin' }
    );
  });

  it('handles API errors gracefully', async () => {
    api.apiClient.getMatchById.mockRejectedValue(new Error('Failed to fetch'));

    await act(async () => {
      render(<MatchViewer matchId="match1" onBack={mockOnBack} />);
    });

    await waitFor(() => {
      expect(screen.getByText('Failed to fetch')).toBeInTheDocument();
    });

    const retryButton = screen.getByText('Retry');
    fireEvent.click(retryButton);

    expect(api.apiClient.getMatchById).toHaveBeenCalledTimes(2);
  });

  it('displays different match statuses correctly', async () => {
    const scheduledMatch = {
      ...mockMatch,
      status: 'scheduled',
      homeScore: 0,
      awayScore: 0,
    };

    await act(async () => {
      render(<MatchViewer match={scheduledMatch} onBack={mockOnBack} />);
    });

    expect(screen.getByText('SCHEDULED')).toBeInTheDocument();
  });

  it('shows live minute for live matches', async () => {
    const liveMatch = {
      ...mockMatch,
      status: 'live',
      minute: 23,
    };

    await act(async () => {
      render(<MatchViewer match={liveMatch} onBack={mockOnBack} />);
    });

    expect(screen.getByText("23'")).toBeInTheDocument();
  });

  it('toggles event list visibility', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Event Timeline'));

    const showMoreButton = screen.getByText('Show more');
    fireEvent.click(showMoreButton);

    expect(screen.getByText('Show less')).toBeInTheDocument();

    fireEvent.click(screen.getByText('Show less'));
    expect(screen.getByText('Show more')).toBeInTheDocument();
  });

  it('formats dates correctly', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    // Check if date is formatted (format depends on locale)
    expect(screen.getByText(/Monday/)).toBeInTheDocument();
    expect(screen.getByText(/January/)).toBeInTheDocument();
  });

  it('displays team crests', async () => {
    await act(async () => {
      render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
    });

    const crests = screen.getAllByRole('img');
    expect(crests[0]).toHaveAttribute('src', '/arsenal.png');
    expect(crests[1]).toHaveAttribute('src', '/chelsea.png');
  });

  it('shows empty events state', async () => {
    const matchWithoutEvents = {
      ...mockMatch,
      events: [],
    };

    await act(async () => {
      render(<MatchViewer match={matchWithoutEvents} onBack={mockOnBack} />);
    });

    fireEvent.click(screen.getByText('Event Timeline'));
    fireEvent.click(screen.getByText('Show more'));

    expect(screen.getByText('No events available for this match')).toBeInTheDocument();
  });

  it('matches snapshot with match data', async () => {
    await act(async () => {
      const { container } = render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
      expect(container).toMatchSnapshot();
    });
  });

  it('matches snapshot in loading state', () => {
    const { container } = render(<MatchViewer matchId="match1" onBack={mockOnBack} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot with admin access', async () => {
    ClerkReact.useUser.mockReturnValue({ user: mockAdminUser });

    await act(async () => {
      const { container } = render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Event type canonicalization', () => {
    it('handles different event type formats', async () => {
      const matchWithVariedEvents = {
        ...mockMatch,
        events: [
          { type: 'GOAL', minute: 10, team: 'Arsenal', player: 'Player 1' },
          { type: 'yellow', minute: 20, team: 'Chelsea', player: 'Player 2' },
          { type: 'sub', minute: 30, team: 'Arsenal', player: 'Player 3' },
        ],
      };

      await act(async () => {
        render(<MatchViewer match={matchWithVariedEvents} onBack={mockOnBack} />);
      });

      fireEvent.click(screen.getByText('Event Timeline'));
      fireEvent.click(screen.getByText('Show more'));

      // Should handle different event type formats gracefully
      expect(screen.getByText(/Goal/)).toBeInTheDocument();
      expect(screen.getByText(/Yellow Card/)).toBeInTheDocument();
      expect(screen.getByText(/Substitution/)).toBeInTheDocument();
    });
  });

  describe('Live commentary section', () => {
    it('renders live commentary component', async () => {
      await act(async () => {
        render(<MatchViewer match={mockMatch} onBack={mockOnBack} />);
      });

      fireEvent.click(screen.getByText('Live Commentary'));

      expect(screen.getByTestId('live-commentary')).toBeInTheDocument();
      expect(screen.getByText(`Live Commentary for match ${mockMatch.id}`)).toBeInTheDocument();
    });
  });
});