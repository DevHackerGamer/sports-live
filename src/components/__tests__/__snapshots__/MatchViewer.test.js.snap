// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchViewer Component matches snapshot in loading state 1`] = `
<div>
  <div
    class="match-viewer"
  >
    <div
      class="loading-overlay"
    >
      <div
        class="loading-spinner"
      />
      <p>
        Loading match details...
      </p>
    </div>
  </div>
</div>
`;

exports[`MatchViewer Component matches snapshot with admin access 1`] = `<div />`;

exports[`MatchViewer Component matches snapshot with match data 1`] = `<div />`;
